python/ops.py

```diff
@@ -63,6 +88,284 @@ def c2r_ifft2(x_fft):
+      tf.cast(nbins, tf.float32)), tf.int32)
+    vb = tf.cast(tf.math.floormod(
+      tf.round((uv[:, 1] - first_bin) / bin_size),
+      tf.cast(nbins, tf.float32)), tf.int32)
+    indices = sub2ind(nbins, ub, vb)
+    if i == 0:
+      N = tf.expand_dims(tf.cast(tf.reshape(tf.math.bincount(
+        indices, minlength=nbins * nbins), [nbins, nbins, 1]),
+        dtype=tf.float32), axis=0)
+      N = N / tf.math.maximum(EPS, tf.reduce_sum(N))
+    else:
+      n = tf.expand_dims(tf.cast(tf.reshape(tf.math.bincount(
+        indices, minlength=nbins * nbins), [nbins, nbins, 1]),
+        dtype=tf.float32), axis=0)
+      n = n / tf.math.maximum(EPS, tf.reduce_sum(n))
+      N = tf.concat((N, n), axis=0)
+  return N
+
+
+def featurize_image(rgb, params):
+  """Produces 2D histograms of a given rgb image.
+
+  This functions produces a 2D histogram of the log-chroma of the input
+  image's colors and the edge image's colors.
+
+  Args:
+    rgb: RGB image (float32) with the shape of [batch_size, height, width,
+      channels].
+    params: a dict with keys:
+    'first_bin': (float) location of the edge of the first histogram bin.
+    'bin_size': (float) size of each histogram bin.
+    'nbins': (int) number of histogram bins.
+
+
+  Returns:
+    chroma_histograms: stack of 2D chroma histograms (float32) from the
+    filter bank. For each channel, the chroma histogram is generated from the
+    input as described below:
+      ch = 0: from RGB input.
+      ch = 1: from edge filter input.
+  """
+
+  N = compute_chroma_histogram(rgb, params)
+
+  edge_rgb = local_absolute_deviation(rgb)
+  N_e = compute_chroma_histogram(edge_rgb, params)
+  chroma_histograms = tf.concat((N, N_e), axis=3)
+
+  return chroma_histograms
+
+
+def local_absolute_deviation(rgb):
+  """Compute a Local Absolute Deviation in sliding 3x3 window fashion.
+
+  This functions computes the edge signal of a given rgb image.
+
+  Args:
+    rgb: RGB image (float32) with the shape of [batch_size, height, width,
+      channels].
+
+  Returns:
+    rgb_edge: RGB image (float32) with the shape of [batch_size, height, width,
+    channels] of the corresponding edge signals of each color channel in the rgb
+    input image.
+  """
+
+  # Padding the image and the mask (if given) before computing the image's
+  # edges
+  for c in range(3):  # for each color channel
+    if c == 0:
+      rgb_padded = tf.expand_dims(
+        tf.pad(rgb[:, :, :, c], [[0, 0], [1, 1], [1, 1]],
+               'SYMMETRIC'), axis=3)
+    else:
+      rgb_padded = tf.concat((rgb_padded,
+                              tf.expand_dims(tf.pad(
+                                rgb[:, :, :, c],
+                                [[0, 0], [1, 1], [1, 1]],
+                                'SYMMETRIC'), axis=3)), axis=3)
+
+  # Applying a serious of conv filters to compute the absolute deviation in
+  # the input image
+
+  for c in range(3):  # for each color channel
```

Comment:
same thing re: for loops

Reply:
Fixed by changing the kernel to apply it once.